design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,229 @@
+pc-q35-rhel9.2.0     RHEL-9.2.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel9.0.0     RHEL-9.0.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel8.6.0     RHEL-8.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.5.0     RHEL-8.5.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.4.0     RHEL-8.4.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.3.0     RHEL-8.3.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.2.0     RHEL-8.2.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.1.0     RHEL-8.1.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.0.0     RHEL-8.0.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel7.6.0     RHEL-7.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+none                 empty machine
+```
+This means each workload must be updated to use a `pc-q35-rhel9.x.x` machine type.
+
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
+	* alternatively, if currently running, take VM offline immediately and update machine type
+	* if not currently running, update machine type immediately
+* Allow user to specify certain VMs for updating machine type automation (e.g. by namespace)
+
+## Non Goals
+While these additions could be beneficial to cluster-admins who wish to use the mass machine type transition (MMTT), these are not covered in the scope of this design proposal: 
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Allowing user to filter VMs in ways other than by namespace; e.g. with label-selector
+
+Also not covered in this design proposal is the method in which users/cluster admins will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with CentOS Stream X without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. Both the minimum supported machine type and the target machine type are constants that can be updated as future versions are released and more machine types are deprecated upon these releases. 
```

Comment:
It's good to look at the "mass" case, but what about a more narrow one, updating one vm?

I did like Felix comment that what about simply making it a direct functionality of virtctl instead of virtctl creating  Job.
Will the virtctl command for creating the job also have a way to cancel the job?

Reply:
> It's good to look at the "mass" case, but what about a more narrow one, updating one vm?
>
> I did like Felix comment that what about simply making it a direct functionality of virtctl instead of virtctl creating Job.

Firstly, I think that the "mass" case is more common, as usually an admin would convert machine types only if he has to, and usually before upgrades. But as for the the case of a single VM (or a few VMs), I think it's still convenient that you have an object that reflects the status of the conversion, and that makes this whole operation more declarative.

Here's another use-case: let's say that an admin converts the machine types of some VMs, then does so for a few more VMs, and so on. Tracking this without jobs would be messy IMO. With jobs, you know exactly which groups of VMs were aimed for conversion, when, and what's the status of the conversion.

In general, if we look at it from a high-level perspective, there are 3 phases to the conversion:
1. Listing the VMs that we want to convert
2. Waiting for them to shut down / restart
3. "Finalizing" the conversion, or letting the user know that the conversion is done

I think that performing steps 2-3 imperatively would be a burden for the administrator.

If we would go with the imperative virtctl-only approach, how would an administrator be able to track progress in an easy way? How could he be notified when the conversion is done, or get a clear status of the operation's state? How would the administrator perform re-tries is something goes wrong? I'm open to discuss this front if we overcome these challenges.

cc @0xFelix 

> Will the virtctl command for creating the job also have a way to cancel the job?

Great question.
Perhaps to make this a whole solution we need something like the following commands:
1. `virtctl machine-type-convert start` - to define the target group of VMs, desired machine type, etc
2. `virtctl machine-type-convert list` - to list the currently running jobs
3. `virtctl machine-type-convert delete` - to delete a job.

Reply:
I did not mean to say that we do not need a mass tool at all.

However, I'm missing 1 critical step (and 1 use-case):
How does the admin try out the migration on one or so VMs before applying it to many more?
If I was an admin, I want to try the migration on all the different VM types I have, before changing all of them.
After all this operation CAN lead to functional regressions, this is the reason why we do not do this automatically.

Thus to me a 2 layered approch just looked reasonable, i.e.
1. `virtctl vm update machine-type $VMNAME` # first trial
2. `virtctl vm update machine-type --jobs $LABELSELECTOR` # now all

Reply:
@fabiand Are you thinking of a hybrid approach where 1 is running locally in `virtctl` while 2 is running as a `Job` on the cluster? For the modification of a single VM the `virtctl` command could be quite trivial, but afterall this would also be covered by @iholder101's suggestion to define the target group of VMs (could also be a single VM). In this case I'd still vote for only the `Job` because this avoids duplicating the updater logic in two places.

To get back to @iholder101's concern about monitoring the state of the operation. I have to agree that it would be easier to track the state of it in a `Job`. How else would you know which VMs were already upgraded? Where do you store the information?

Reply:
I do think that in the case of a single VM, we could still utilize the same implementation method of invoking the `Job` on that single VM, but allowing the user to specify that VM by name. I currently have this listed as a follow up in the implementation phases section of the design proposal, but if you feel this would be important in the initial design I can also add this single VM use case to the initial design and implement a flag to specify a VM by name. I agree with @0xFelix that I would like to avoid duplicating the logic for updating the VMs and keeping the design limited to the `Job` approach.